src/Java.Interop/Java.Interop/JniNativeMethodRegistrationArguments.cs

```diff
@@ -0,0 +1,22 @@
+﻿using System.Collections.Generic;
+
+namespace Java.Interop
+{
+	public struct JniNativeMethodRegistrationArguments
+	{
+		public ICollection<JniNativeMethodRegistration> Registrations { get { return _registrations; } }
+		public string Methods { get; }
+		List<JniNativeMethodRegistration> _registrations;
+
+		public JniNativeMethodRegistrationArguments (List<JniNativeMethodRegistration> registrations, string methods)
+		{
+			_registrations = registrations;
+			Methods = methods;
+		}
+
+		public void AddRange (IEnumerable<JniNativeMethodRegistration> registrations)
+		{
+			_registrations.AddRange (registrations);
```

Comment:
This type is a `struct`. As such, it has an implicit default constructor, from which `_registrations` will be `null`.

We should choose one of three things:

1. Ignore it, and realize that `NullReferenceException` can be thrown from here
2. `null`-check `_registrations`, and throw `InvalidOperationException` when `null`.
3. Use `_registrations?.AddRange()`, and "ignore" it.

I'm inclined toward (3).

Reply:
Additionally, an alternate idea comes to mind:

If the constructor instead took `ICollection<JniNativeMethodRegistration>`, we could support that by just using `.Add()` here:

```csharp
foreach (var v in registrations) {
    Registrations.Add (v);
}
```

This would allow removing the Gendarme ignore on the constructor, and would allow non-`List<JniNativeMethodRegistration>` values to be provided. It just means we have more `.Add()` calls.

Reply:
That is how I implemented it first. Later I used List::AddRange to improve performance, which I think is our goal here.

Optionally we can use `ICollection<JniNativeMethodRegistration>` and try to cast it to `List<JniNativeMethodRegistration>` and use `AddRange` in that case?

Reply:
I think it would help "intrinsic understanding" ("who reads documentation?") if the constructor parameter type matched the property type.

As such, both the constructor parameter and the property should either be `ICollection<T>` or `List<T>`. (Not sure how this would impact Gendarme, but for *consistency and understanding* they should be the same type.)

Given that we're mostly after performance, I think using `List<T>` everywhere would be better.

We still need to handle/answer the question of "what happens when `Registrations` is `null`?"

Reply:
OK, I changed them both to `ICollection<T>` and in AddRegistrations I check whether it can be cast to `List<T>` to speed it up.

When `Registrations` is null, I made it throw an `InvalidOperationException` as it is invalid state we would like to avoid. We should not get into such state, so throwing the exception might potentially helps us find a bug.